File: Chloe Gwon/2022-07-06-House-Robber-II.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    if (nums.length === 1) return nums[0];
    return Math.max(robHouse(nums, 0, nums.length-2),
                   robHouse(nums, 1, nums.length-1));
};

function robHouse(nums, start, end){
    var prev1 = 0;
    var prev2 = 0;
    for (var i=start; i<=end; i++){
        var temp = prev1;
        prev1 = Math.max(prev1, prev2 + nums[i]);
        prev2 = temp;
    }
    return prev1;
}

File: Chris/2022-07-06-HouseRobberII.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        
        if(len(nums) < 4):
            return max(nums)
        
        def robArray(arrayNums: List[int]) -> int:
            
            dp = [0 for i in range(len(arrayNums) + 2)]
            
            for i in range(len(arrayNums)-1, -1, -1):
                dp[i] = max(dp[i+1], dp[i+2] + arrayNums[i])
                
            return dp[0]
        
        
        return max(robArray(nums[1:]) , robArray(nums[2:-1]) + nums[0])